source/common/stats/allocator_impl.cc

```diff
@@ -78,27 +68,55 @@ template <class BaseClass> class StatsSharedImpl : public MetricImpl<BaseClass>
   // RefcountInterface
   void incRefCount() override { ++ref_count_; }
   bool decRefCount() override {
+    // We must, unfortunately, hold the allocator's lock when decrementing the
+    // refcount. Otherwise another thread may simultaneously try to allocate the
+    // same name'd stat after we decrement it, and we'll wind up with a
+    // dtor/update race. To avoid this we must hold the lock until the stat is
+    // removed from the map.
+    //
+    // It might be worth thinking about a race-free way to decrement ref-counts
+    // without a lock, for the case where ref_count > 2, and we don't need to
+    // destruct anything. But it seems preferable at to be conservative here,
+    // as stats will only go out of scope when a scope is destructed (during
+    // xDS) or during admin stats operations.
+    Thread::LockGuard lock(alloc_.mutex_);
     ASSERT(ref_count_ >= 1);
-    return --ref_count_ == 0;
+    if (--ref_count_ == 0) {
+      alloc_.sync().syncPoint(AllocatorImpl::DecrementToZeroSyncPoint);
+      removeFromSetLockHeld();
+      return true;
+    }
+    return false;
   }
   uint32_t use_count() const override { return ref_count_; }
 
+  /**
+   * We must atomically remove the counter/gauges from the allocator's sets when
+   * our ref-count decrement hits zero. The counters and gauges are held in
+   * distinct sets so we virtualize this removal helper.
+   */
+  virtual void removeFromSetLockHeld() EXCLUSIVE_LOCKS_REQUIRED(alloc_.mutex_) PURE;
+
 protected:
   AllocatorImpl& alloc_;
 
   // Holds backing store shared by both CounterImpl and GaugeImpl. CounterImpl
   // adds another field, pending_increment_, that is not used in Gauge.
   std::atomic<uint64_t> value_{0};
+  std::atomic<uint32_t> ref_count_{0};
```

Comment:
Technically we don't need the atomic anymore since we are under lock, right? For clarity WDYT about removing the atomic and adding a comment here on how we are synchronizing? Should we also add an ABSL locking notation around ref_count_ for now?

Reply:
We could drop the atomic but right now a copy or assign will cause a ref-count increment without a lock. We'd need to take locks for those operations too.

Reply:
OK makes sense. Can you just add some comments here then about the locking around the ref count?